hir/src/ir/traits/types.rs

```diff
@@ -68,12 +68,67 @@ derive! {
 }
 
 derive! {
-    /// Op expects all operands and results to be of the same type
+    /// Op expects all operands and results to be of the same type.
+    /// NOTE: Operations that implements this trait must also explicitely implement
+    /// [`SameTypeOperands`]. This can be achieved by using the "traits" filed in the [#operation]
+    /// macro.
+    /// Like so:
     ///
-    /// TODO(pauls): Implement verification for this. Ideally we could require `SameTypeOperands`
-    /// as a super trait, check the operands using its implementation, and then check the results
-    /// separately
-    pub trait SameOperandsAndResultType {}
+    /// #[operation (
+    ///         dialect = ArithDialect,
+    ///         traits(UnaryOp, SameTypeOperands, SameOperandsAndResultType),
+    ///         implements(InferTypeOpInterface, MemoryEffectOpInterface)
+    /// )]
+    /// pub struct SomeOp {
+    ///       (...)
+    /// }
+    pub trait SameOperandsAndResultType: SameTypeOperands {}
+
+    verify {
+        fn operands_and_result_are_the_same_type(op: &Operation, context: &Context) -> Result<(), Report> {
+            let mut operands = op.operands().iter();
+            if let Some(first_operand) = operands.next() {
+                let (expected_ty, set_by) = {
+                    let operand = first_operand.borrow();
+                    let value = operand.value();
+                    (value.ty().clone(), value.span())
+                };
+
+                let results = op.results().iter();
+
+                for result in results {
+                    let result = result.borrow();
+                    let value = result.as_value_ref().borrow();
+                    let result_ty = result.ty();
+
+                    if result_ty != &expected_ty {
+                        return Err(context
+                            .session()
+                            .diagnostics
+                            .diagnostic(Severity::Error)
+                            .with_message(::alloc::format!("invalid operation result {}", op.name()))
+                            .with_primary_label(
+                                op.span,
+                                "this operation expects the operands and the results to be of the same type"
+                            )
+                            .with_secondary_label(
+                                set_by,
+                                "inferred the expected type from this value"
+                            )
+                            .with_secondary_label(
+                                value.span(),
+                                "which differs from this value"
```

Comment:
nit: Should this be something more like the following:
```suggestion
                                "which differs from this value's type"
```

Reply:
Fair point!. I tried to replicate the structure/text present in the original verifier here: https://github.com/0xMiden/compiler/blob/f0ff05aca81e9dc579899223d2f19bb7200fbffd/hir/src/ir/traits/types.rs#L40-L61.

But I do think that it is clearer with your suggestion.